source/computer/shell.rst

```diff
@@ -0,0 +1,284 @@
+Shell 基础
+===========
+
+:本节贡献者: |姚家园|\（作者）、
+             |田冬冬|\（审稿）
+:最近更新日期: 2021-02-13
+:预计花费时间: 120 分钟
+
+----
+
+扩展
+-----
+
+Shell 接收到用户输入的命令以后，会根据空格将用户的输入，拆分成一个个词元（token）。
+如果词元中存在特殊字符，Shell 会先扩展这些特殊字符。扩展完成后，Shell 才会把命令交给系统执行。
+需要注意，\ **扩展是由 Shell 负责的，与所执行的命令无关**\ 。命令本身并不扩展参数，收到什么参数就原样执行。
+本文只介绍 Shell 扩展的一些常见用法，可以参考
+`Bash 的模式扩展 <https://wangdoc.com/bash/expansion.html#startend-%E6%89%A9%E5%B1%95>`__
+学习详细用法。
+
+例如，用户在终端键入 ``cd ~`` 后，Shell 先将该命令拆分成 ``cd`` 和 ``~`` 两个词元，
+然后把 ``~`` 这个特殊字符扩展成用户的家目录（如 ``/home/seismo-learn``\ ），最后交给系统执行，
+即 ``cd /home/seismo-learn``。系统执行完命令后，用户所在目录就切换到家目录下了。
+
+.. note::
+
+   Shell 扩展在网上有很多教程和说明，但是一些名词的定义和说法不一。比如有人将扩展称为模式扩展，
+   有人将扩展的英文写为 globbing 并称通配符叫 globbing patterns，有人将扩展和通配符等价，
+   有人将扩展特指文件名扩展，有人将通配符特指为 ``?``\ 、\ ``*``\ 以及\ ``[]``\ 等等。
+
+   笔者使用 ``man sh`` 命令在 CentOS 7 系统下查看了 Bash 这一常见 Shell 的手册，并未搜索到
+   wildcards 和 globbing 这两个单词。可能由于多种原因（历史原因、不同程序原因等），这些名词被保留下来
+   并常使用。在本教程中，我们不想（其实也是不能）区分这些名词的详细定义，而是参考
+   `GNU Bash Reference Manual: 3.5 Shell Expansions <https://www.gnu.org/savannah-checkouts/gnu/bash/manual/bash.html#Shell-Expansions>`__
+   直接使用“扩展”这一词语指代 Shell 扩展特殊字符的全部过程，并将这些特殊字符都广义地称为通配符（wildcards）。
+   毕竟日常科研中，我们主要是使用这些功能，而不是去深究这些功能背后的历史和定义。
+   但是需要注意，这些特殊字符作用的顺序是不同的；狭义的通配符一般可能只指 ``?``\ 、\ ``*``\ 以及\ ``[]``\ 。
+
+.. table:: Bash 常见特殊字符（本教程也称为通配符）
+   :align: center
+
+   =============== ===================================================== ============
+   特殊字符	        作用                                                   例子
+   =============== ===================================================== ============
+   ~               匹配当前用户的家目录                                     例如 /home/seismo-learn
+   ?	           匹配任意单个字符（不能匹配零个字符）                       ?.txt 匹配 a.txt、b.txt 等文件名
+   \*	           匹配任意数量字符（能匹配零个字符）                         \*.txt 匹配 a.txt、b.txt、ab.txt 等文件名
+   [...]           匹配方括号中的任意一个字符                               [abc].txt 匹配 a.txt、b.txt 以及 c.txt 文件名
+   [start-end]     方括号扩展的简写模式，匹配一个连续的范围                   [a-z] 表示所有小写字母，[0-9] 等同于 [0123456789]
+   [^...]          匹配不在方括号里面的任意一个字符                          [^ab]c.txt 匹配 cc.txt、dc.txt 等文件名
+   [!...]          与 [^...] 等价                                         [!ab]c.txt 匹配 cc.txt、dc.txt 等文件名
+   {...}           扩展大括号里面的所有值，各个值之间使用逗号分隔              s{e,i,s}m 扩展成 sem sim ssm
+   {start..end}    大括号扩展的简写模式，扩展一个连续的范围                   s{a..c}m 扩展成 sam sbm scm
```

Comment:
Do you use these expansions in your daily research? If not, we may just remove them.

Reply:
`~`、`*` are used in daily research.

Reply:
We may remove `{}` and keep `[]`?

Reply:
Yes, maybe keep the first 6 symbols and remove the last three.

Reply:
Done in 8dd06cf.